snap/integrity/integrity_test.go

```diff
@@ -0,0 +1,170 @@
+		expectedOutput uint64
+	}{
+		{0, 0},
+		{1, blockSize},
+		{blockSize, blockSize},
+		{blockSize + 1, 2 * blockSize},
+	} {
+		ret := align(tc.input)
+		c.Check(ret, Equals, tc.expectedOutput, Commentf("%v", tc))
+	}
+}
+
+func (s *IntegrityTestSuite) TestIntegrityHeaderMarshalJSON(c *C) {
+	dmVerityBlock := &dmverity.Info{}
+	integrityDataHeader := integrity.NewIntegrityDataHeader(dmVerityBlock, 4096)
+
+	jsonHeader, err := json.Marshal(integrityDataHeader)
+	c.Assert(err, IsNil)
+
+	c.Check(json.Valid(jsonHeader), Equals, true)
+}
+
+func (s *IntegrityTestSuite) TestIntegrityHeaderUnmarshalJSON(c *C) {
+	var integrityDataHeader integrity.IntegrityDataHeader
+	integrityHeaderJSON := `{
+		"type": "integrity",
+		"size": "4096",
+		"dm-verity": {
+			"root-hash": "00000000000000000000000000000000"
+		}
+	}`
+
+	err := json.Unmarshal([]byte(integrityHeaderJSON), &integrityDataHeader)
+	c.Assert(err, IsNil)
+
+	c.Check(integrityDataHeader.Type, Equals, "integrity")
+	c.Check(integrityDataHeader.Size, Equals, uint64(4096))
+	c.Check(integrityDataHeader.DmVerity.RootHash, Equals, "00000000000000000000000000000000")
+}
+
+func (s *IntegrityTestSuite) TestIntegrityHeaderEncode(c *C) {
+	var integrityDataHeader integrity.IntegrityDataHeader
+	magic := integrity.Magic
+
+	header, err := integrityDataHeader.Encode()
+	c.Assert(err, IsNil)
+
+	magicRead := header[0:len(magic)]
+	c.Check(magicRead, DeepEquals, magic)
+
+	nullByte := header[len(header)-1:]
+	c.Check(nullByte, DeepEquals, []byte{0x0})
+
+	c.Check(uint64(len(header)), Equals, integrity.Align(uint64(len(header))))
+}
+
+func (s *IntegrityTestSuite) TestIntegrityHeaderUnserialize(c *C) {
+	var integrityDataHeader integrity.IntegrityDataHeader
+	magic := integrity.Magic
+
+	integrityHeaderJSON := `{
+		"type": "integrity",
+		"size": "4096",
+		"dm-verity": {
+			"root-hash": "00000000000000000000000000000000"
+		}
+	}`
+	header := append(magic, integrityHeaderJSON...)
+	header = append(header, 0)
+
+	headerBlock := make([]byte, 4096)
+	copy(headerBlock, header)
+
+	err := integrityDataHeader.Decode(headerBlock)
+	c.Assert(err, IsNil)
+
+	c.Check(integrityDataHeader.Type, Equals, "integrity")
+	c.Check(integrityDataHeader.Size, Equals, uint64(4096))
+	c.Check(integrityDataHeader.DmVerity.RootHash, Equals, "00000000000000000000000000000000")
+}
+
+func (s *IntegrityTestSuite) TestGenerateAndAppendSuccess(c *C) {
+	blockSize := uint64(integrity.BlockSize)
+
+	snapPath, _ := snaptest.MakeTestSnapInfoWithFiles(c, "name: foo\nversion: 1.0", nil, nil)
+
+	snapFileInfo, err := os.Stat(snapPath)
+	c.Assert(err, IsNil)
+	orig_size := snapFileInfo.Size()
+
+	err = integrity.GenerateAndAppend(snapPath)
+	c.Assert(err, IsNil)
+
+	snapFile, err := os.Open(snapPath)
+	c.Assert(err, IsNil)
+	defer snapFile.Close()
+
+	// check integrity header
+	_, err = snapFile.Seek(orig_size, io.SeekStart)
+	c.Assert(err, IsNil)
+
+	header := make([]byte, blockSize-1)
+	n, err := snapFile.Read(header)
```

Comment:
This is fixed and changed to simply `HeaderSize` and pushed in one of the later branches. 

context: In the end I opted for making one more const `HeaderSize` same size as `blockSize`. While this may not be the most "dynamic" approach, it simplified several calculations as the current header cannot in reality exceed that (it only contains a fixed-size string "integrity", a size of uint64 and the dm-verity's root hash). That -1 was indeed wrong here as I was trying to account for the extra null byte while I shouldn't.

Reply:
It is a bit strange as I do cannot find the happy case (TestIntegrityHeaderEncode()) anymore?